### DISCUSSION TITLE
Trouble prefix

### DISCUSSION THREAD

**Discussion by chichako** [audacious-media-player/audacious#1510] [Q&A]
Debian 12, audacious-4.4.2, audacious-plugins-4.4.2
When compiling with ./configure --prefix=/opt/audacious I get an error

`audacious -V
INFO dbus-server.cc:1132 [name_acquired]: Owned D-Bus name (org.atheme.audacious) on session bus.
INFO main.cc:399 [main]: No remote session; starting up.
INFO plugin-init.cc:140 [start_required]: Probing for output plugin.
ERROR plugin-init.cc:155 [start_required]: No output plugin found.
(Did you forget to install audacious-plugins?)`

Without the prefix (./configure) everything works fine.
I add environment variable for plugins

`export PKG_CONFIG_PATH=/opt/audacious/lib/pkgconfig:$PKG_CONFIG_PATH`

to get rid of

`configure: error: Cannot find Audacious 4.4; have you installed Audacious yet?`

What am I doing wrong or is this a bug?

P.S. Excuse my English! (Google Translate)

**Comment by mschwendt** [comment]
What do you get when running "**pkg-config --variable=plugin_dir audacious**"?

After compiling and installing the "audacious" package, have you built and installed the "audacious-plugins" package, so the plugins are located within that directory as specified by pkg-config?

**Comment by chichako** [comment]
`What do you get when running "pkg-config --variable=plugin_dir audacious"?`

_/usr/lib/x86_64-linux-gnu/audacious_

`After compiling and installing the "audacious" package, have you built and installed the "audacious-plugins" package, so the plugins are located within that directory as specified by pkg-config?`

No. Why? I installed the application and plugins in /opt/audacious (--prefix=/opt/audacious) They should logically be in /opt/audacious/lib

**Comment by mschwendt** [comment]
Since pkg-config tells a different installation path for the plugins, you've made a mistake somewhere. It looks like you've kept a second installation of Audacious below /usr. That can lead to confusion.

The reason you need to "export PKG_CONFIG_PATH=..." is that pkg-config doesn't know about the custom path /opt/audacious/lib/pkgconfig you want use. You can verify the audacious.pc file contents in that directory and check whether it defines correct variables. You also need to set that PKG_CONFIG_PATH when building the plugins. Have you done that?

What files are available below /opt/audacious you haven't shown.

**Comment by chichako** [comment]
`It looks like you've kept a second installation of Audacious below /usr`
No, this is a clean virtual machine. There were no other installations from the repository or compilations
`You also need to set that PKG_CONFIG_PATH when building the plugins. Have you done that?`
Yes, I did. Before "configure" I did "export PKG_CONFIG_PATH=/opt/audacious/lib/pkgconfig:$PKG_CONFIG_PATH"

**/opt/audacious/lib/pkgconfig**

prefix=/opt/audacious
exec_prefix=${prefix}
datarootdir=${prefix}/share
data_dir=${datarootdir}/audacious

version=4.4.2
include_dir=${prefix}/include
audacious_include_dir=${prefix}/include
lib_dir=${exec_prefix}/lib

plugin_dir=${exec_prefix}/lib/audacious

Name: Audacious
Description: Audacious is a versatile and handy multi platform media player
Version: 4.4.2
Libs: -L${lib_dir} -laudcore
Cflags: -I${audacious_include_dir}

`What files are available below /opt/audacious you haven't shown.`
[pastebin](https://pastebin.com/8WbYTnmx)

**Reply by mschwendt** [reply]
Earlier you've shown the output of "pkg-config --variable=plugin_dir audacious", and it didn't point at your plugin dir below /opt/audacious but it printed /usr/lib/x86_64-linux-gnu/audacious. That is proof you have an audacious.pc file in default pkgconfig path, which is about an installation NOT in /opt/audacious. It indicates that you've make some mistake while building Audacious. Running "pkg-config --path audacious" would tell where that audacious.pc file is located which you've accessed earlier. Running "which audacious" would tell whether you really execute /opt/audacious/bin/audacious and not a copy in a different path.

Maybe it's time to run "strace $(which audacious) &> output.txt" and attach the output.

**Reply by chichako** [reply]
> Earlier you've shown the output of "pkg-config --variable=plugin_dir audacious", and it didn't point at your plugin dir below /opt/audacious but it printed /usr/lib/x86_64-linux-gnu/audacious. That is proof you have an audacious.pc file in default pkgconfig path, which is about an installation NOT in /opt/audacious. It indicates that you've make some mistake while building Audacious. Running "pkg-config --path audacious" would tell where that audacious.pc file is located which you've accessed earlier. Running "which audacious" would tell whether you really execute /opt/audacious/bin/audacious and not a copy in a different path.
> 
> Maybe it's time to run "strace $(which audacious) &> output.txt" and attach the output.

Sorry! You were right. I forgot to change the virtual machine from the backup. There really were previous installations. I will publish the result of "strace", but first check if I did everything correctly

**Audacious:**

```
cd ~/audacious
./autogen.sh
./configure --prefix=/opt/audacious
make -j 4
sudo make install
```

**Plugins:**

```
cd ~/audacious-plugins
./autogen.sh
export PKG_CONFIG_PATH=/opt/audacious/lib/pkgconfig:$PKG_CONFIG_PATH
./configure --prefix=/opt/audacious
make -j 4
sudo make install
```

**Comment by chichako** [comment]
strace
[output.txt](https://github.com/user-attachments/files/17700759/output.txt)

**Comment by chichako** [comment]
`pkg-config --variable=plugin_dir audacious`
Returns nothing. It looks like the PKG_CONFIG_PATH export was not applied.

**Reply by mschwendt** [reply]
Well, for an environment variable to be set always, you would need to export it within your shell startup files such as ~/.bash_profile or whatever shell you're using.

The strace output refers to a different installation path this time: /opt/audacious/lib/x86_64-linux-gnu/ compared with the earlier /opt/audacious/lib/

I assume the audacious.pc file also ended up in the pkgconfig directory below that, so what you've specified as PKG_CONFIG_PATH likely was wrong when building the plugins. Therefore, according to the end of the strace, no plugins were found.

**Reply by chichako** [reply]
> I assume the audacious.pc file also ended up in the pkgconfig directory below that, so what you've specified as PKG_CONFIG_PATH likely was wrong when building the plugins. Therefore, according to the end of the strace, no plugins were found.

What should I put in $PKG_CONFIG_PATH so that the program runs correctly from /opt/audacious? Write how to configure it correctly

**Reply by mschwendt** [reply]
pkg-config is completely unrelated to running Audacious. It is a build-time tool. 

Since you override $prefix when you run "configure", you need to check where the "audacious.pc" file for pkg-config is installed. So far, the paths you've quoted differ. Based on your recent comments, I assume it is /opt/audacious/lib/x86_64-linux-gnu/pkgconfig with the arch-specific directory in there. When you build "audacious-plugins", the plugin directory will be retrieved from the audacious.pc file by pkg-config, so as a test I've suggested "pkg-config --variable=plugin_dir audacious" to check whether it runs successfully and prints the proper path. If it doesn't, you've not modified $PKG_CONFIG_PATH correctly.

**Reply by chichako** [reply]
> pkg-config is completely unrelated to running Audacious. It is a build-time tool.
> 
> Since you override $prefix when you run "configure", you need to check where the "audacious.pc" file for pkg-config is installed. So far, the paths you've quoted differ. Based on your recent comments, I assume it is /opt/audacious/lib/x86_64-linux-gnu/pkgconfig with the arch-specific directory in there. When you build "audacious-plugins", the plugin directory will be retrieved from the audacious.pc file by pkg-config, so as a test I've suggested "pkg-config --variable=plugin_dir audacious" to check whether it runs successfully and prints the proper path. If it doesn't, you've not modified $PKG_CONFIG_PATH correctly.

The *pc file is in the "/opt/audacious/lib/pkgconfig" folder. I specify it when exporting the "$PKG_CONFIG_PATH" environment variable. Before that, there was one error due to previous compilations. I don't understand why it adds **x86_64-linux-gnu** to the "/opt/audacious/lib/x86_64-linux-gnu/audacious/Container" plugin path. When they are in "/opt/audacious/lib/audacious". I think that this is somehow related to the initial default configuration during compilation and to /etc/ld.so.conf.d/ that it uses. When configuring, you also need to specify some additional flags and keys for compilation when changing the prefix, then it will work correctly, but I have no experience in this matter.

**Reply by mschwendt** [reply]
> I don't understand why it adds **x86_64-linux-gnu**

It does that because your installation is a multi-arch environment. And thus you get arch-specific paths like /usr/lib/x86_64-linux-gnu/ instead of just /usr/lib/. With a different distribution it might be /usr/lib64 versus /usr/lib, /usr/lib64/pkgconfig versus /usr/lib/pkgconfig, and so on.

Btw, pkgconfig is unrelated to ld.so.conf as it is a build-time tool whereas ld.so is about libraries at runtime.

**Comment by chichako** [comment]
Found the reason, in addition to `PKG_CONFIG_PATH` it was necessary to put another variable in `.profile` or `*.desktop` `LD_LIBRARY_PATH`. In my case it is `export LD_LIBRARY_PATH=/opt/audacious/lib:$LD_LIBRARY_PATH`. Thanks to everyone for the answers!

**Reply by mschwendt** [reply]
That indicates you have not configured ld.so (the runtime dynamic linker) correctly in e.g. /etc/ld.so.conf.d/ for your customized $prefix.

Recommended reading: https://www.hpc.dtu.dk/?page_id=1180
